include/pybind11/detail/internals.h

```diff
@@ -68,6 +68,7 @@ struct internals {
     type_map<type_info *> registered_types_cpp; // std::type_index -> pybind11's type information
     std::unordered_map<PyTypeObject *, std::vector<type_info *>> registered_types_py; // PyTypeObject* -> base type_info(s)
     std::unordered_multimap<const void *, instance*> registered_instances; // void * -> instance*
+    type_map<const std::type_info *> holders_seen; // type -> seen holder type (to detect holder conflicts)
```

Comment:
I was puzzled why you use a global map `holders_seen` here, instead of storing the unique holder type of a registered type [in its typeinfo](https://github.com/pybind/pybind11/pull/2052/files#diff-b47e807c57dacc02a52aa80c712d77051f145748c3be9b8c4e90dd1aa27f7e34R128-L133) as I did in #2052. However, your approach just considers the first occurrence of a holder type as the gold standard. :+1: